### DISCUSSION TITLE
Theseus installation on Google Colab

### DISCUSSION THREAD

**Discussion by saxenam06** [facebookresearch/theseus#347] [Q&A] [answered]
Hello, Do you have any plans to extend the Theseus installation in Google Colab. The main issue are follows which restricts usage of Theseus on colab. 

1. Colab doesn't support docker yet so it's not possible to build the wheels with the build script.
2. Colab cuda version is 11.2 which is not supported by the build script.
3. I tried to change cuda version to 10.2 so as to install Theseus with pip. Though the installation complete successfully But I get error while Importing Theseus: cannot import name '__C' from functorch. . 

I would highly appreciate your thoughts on the same. Many thanks!

**Comment by mhmukadam** [comment, marked as answer]
Hi, you can easily install Theseus using the commands below to run in colab:
```python
!apt install libsuitesparse-dev
!pip install theseus-ai

import torch
import theseus as th

th.rand_se2(2) # to test
```

The cuda version issue happened recently as torch 1.13 stoped supporting cuda 10.2. We are fixing this in #345 and should be merged soon.

**Reply by saxenam06** [reply]
Super, works perfectly. Many thanks for the quick help.

**Reply by saxenam06** [reply]
Hi, During the usage of Theseus (after getting it installed in colab with your help) in one of the repos, I am getting the following error. Is BaSpaCho sparse solver required to be installed separately? Would highly appreciate if you may please share your thoughts again on below error. Many thanks!

Traceback (most recent call last):
  File "train.py", line 11, in <module>
    from model.planner import MotionPlanner
  File "/content/drive/MyDrive/DIPP_AWS/in_data_dipp/DIPP/model/planner.py", line 2, in <module>
    import theseus as th
  File "/content/drive/MyDrive/DIPP_AWS/in_data_dipp/DIPP/theseus/theseus/__init__.py", line 74, in <module>
    from .optimizer.linear import (  # usort: skip
  File "/content/drive/MyDrive/DIPP_AWS/in_data_dipp/DIPP/theseus/theseus/optimizer/linear/__init__.py", line 11, in <module>
    from .baspacho_sparse_solver import BaspachoSparseSolver
  File "/content/drive/MyDrive/DIPP_AWS/in_data_dipp/DIPP/theseus/theseus/optimizer/linear/baspacho_sparse_solver.py", line 12, in <module>
    from theseus.optimizer.autograd import BaspachoSolveFunction
  File "/content/drive/MyDrive/DIPP_AWS/in_data_dipp/DIPP/theseus/theseus/optimizer/autograd/__init__.py", line 8, in <module>
    from .cholmod_sparse_autograd import CholmodSolveFunction
  File "/content/drive/MyDrive/DIPP_AWS/in_data_dipp/DIPP/theseus/theseus/optimizer/autograd/cholmod_sparse_autograd.py", line 7, in <module>
    from sksparse.cholmod import Factor as CholeskyDecomposition
  File "sksparse/cholmod.pyx", line 1, in init sksparse.cholmod
ValueError: numpy.ndarray size changed, may indicate binary incompatibility. Expected 88 from C header, got 80 from PyObject

**Reply by luisenp** [reply]
Hi @saxenam06. This error is related to the CUDA version issue @mhmukadam mentioned above, which also affects our pypi wheels. This issue should be solved when we release 0.1.3, which will happen soon (maybe even today). 

Were you able to change CUDA version in Colab? We are still planning to keep 10.2 as the default for now, but curious to know if this would be an issue.

**Reply by saxenam06** [reply]
Hi @luisenp, Many Thanks for your response. Yes, I was able to change the CUDA version in Colab to 10.2. This is what I get in Colab with !nvcc --version: 
![image](https://user-images.githubusercontent.com/83720464/199987713-fc1b1885-d324-44c9-82f7-87c08e61f231.png)

But after this when I follow the below steps as @mhmukadam mentioned above, I get the error below that Failed building wheels for theseus-ai. So basically the below steps works only with a fresh Colab runtime without any changes done in the CUDA version. Thanks again for your time into this issue!

!apt install libsuitesparse-dev
!pip install theseus-ai
 
![image](https://user-images.githubusercontent.com/83720464/199989068-c600f62f-b310-49c9-aa55-abc7eb8e88a5.png)

**Reply by luisenp** [reply]
Yes, this is expected. Should be fixed in 0.1.3.

**Reply by saxenam06** [reply]
Thanks a lot for your fix. It works now at my end. With CUDA version changed to 10.2 in COLAB and then installing torch, libsuitesparse and Theseus in a Conda Python 3.9 environment, I am able to run my scripts without solver/torch/Theseus build related issues mentioned before. Thanks a lot for your continuous support.

**Reply by luisenp** [reply]
Glad to hear that it works, and happy to help!
